pkg/plugin/plugin.go

```diff
@@ -77,52 +83,63 @@ func NewJiraPlugin(ctx context.Context, cfg *PluginConfig) (*JiraPlugin, error)
 	if got, want := req.Justification.Category, jiraCategory; got != want {
-		return nil,
-			fmt.Errorf("failed to perform validation, expected category %q to be %q",
-				got, want)
+		err := fmt.Errorf("failed to perform validation, expected category %q to be %q", got, want)
+		log.Printf("failed to validate jira justification: %v", err)
+		return standardValidationErrResponse(err.Error()), nil
 	}
 
 	if req.Justification.Value == "" {
-		return nil, fmt.Errorf("empty justification value")
+		err := errors.New("empty justification value")
+		log.Printf("failed to validate jira justification: %v", err)
+		return standardValidationErrResponse(err.Error()), nil
 	}
 
-	result, err := j.validator.MatchIssue(ctx, req.Justification.Value)
+	result, err := j.validateWithJiraEndpoint(ctx, req.Justification.Value)
 	if err != nil {
-		return &jvspb.ValidateJustificationResponse{
-			Valid: false,
-			Error: []string{err.Error()},
-		}, fmt.Errorf("failed to validate justification: %w", err)
+		log.Printf("failed to validate with jira endpoint: %v", err)
+		if errors.Is(err, validationError.ErrInvalidJustification) {
+			return standardValidationErrResponse(
+					fmt.Sprintf("invalid jira justification %q, ensure you input a valid jira id for an open issue", req.Justification.Value)),
+				nil
+		} else {
+			return nil, standardInternalErr(req.Justification.Value)
+		}
 	}
-
-	if len(result.Matches) == 0 || len(result.Matches[0].MatchedIssues) == 0 {
-		return nil, fmt.Errorf("failed to get the matched jira issue %q", req.Justification.Value)
+	issueID := strconv.Itoa(result.MatchedIssues[0])
+	// The format for the Jira issue URL follows the pattern "https://your-domain.atlassian.net/browse/<issueKey>".
+	issueURL, err := url.JoinPath(j.issueBaseURL, "browse", req.Justification.Value)
+	if err != nil {
+		log.Printf("failed to build a clickable url for issue: %v", err)
+		return nil, standardInternalErr(req.Justification.Value)
 	}
 
-	// There is only one JQL and one issueKey, so the first match result is
-	// checked directly.
-	if len(result.Matches[0].MatchedIssues) == 1 {
-		issueID := strconv.Itoa(result.Matches[0].MatchedIssues[0])
+	return &jvspb.ValidateJustificationResponse{
+		Valid:   true,
+		Warning: result.Errors,
+		Annotation: map[string]string{
+			jiraIssueID:  issueID,
+			jiraIssueURL: issueURL,
+		},
+	}, nil
+}
 
-		// The format for the Jira issue URL follows the pattern "https://your-domain.atlassian.net/browse/<issueKey>".
-		issueURL, err := url.JoinPath(j.issueBaseURL, "browse", req.Justification.Value)
-		if err != nil {
-			return nil, fmt.Errorf("failed to build a clickable url for issue %q", req.Justification.Value)
-		}
+// Validate returns the validation result.
+func (j *JiraPlugin) validateWithJiraEndpoint(ctx context.Context, justificationValue string) (*validator.Match, error) {
```

Comment:
Maybe as a followup: I think all the logic in this func should belong to `j.validator.MatchIssue` rather than this struct

Reply:
Done I added a TODO in comment